### PR TITLE
Bugfix: Fix interrupted network read operation

### PR DESCRIPTION
**Pull request by aggarw13**
### Problem

The `MQTT_ProcessLoop` and `MQTT_ReceiveLoop` read incoming MQTT packet payload over the network by calling the `recvExact` function. The `recvExact` function can be called multiple times to read the expected number of bytes for the MQTT packet but it also implements a timeout functionality of receiving the expected number of payload within the timeout value passed to the function.
This causes problems when the `Transport_Recv` call returns less than requested number of bytes, and there is a timeout (for example, when calling `MQTT_ProcessLoop` with 0ms duration) which causes the function to assume failure instead of reading the remaining payload of the MQTT packet by calling `Transport_Recv` again. Thus, in such cases, the MQTT connection is severed prematurely even though there is a high probability of receiving the remaining bytes of the MQTT packet over the network.

### Solution
Instead of implementing a timeout on the entire duration of receiving the expected number of remaining MQTT packet bytes in `recvExact`, the use of timeout is being changed to be relevant only on the total time of receiving 0 bytes over the network over multiple calls to `Transport_Recv`.
As this modified meaning of the timeout duration is now unrelated to the timeout duration that the `MQTT_ProcessLoop` or `MQTT_ReceiveLoop` functions are called, a new configuration constant for the `recvExact` timeout value, `MQTT_RECV_POLLING_TIMEOUT_MS`, has been added to the library which will carry a default value of 10ms.

### CODE REVIEW

File: source/core_mqtt.c [hunk 1]
```diff
@@ -719,8 +726,11 @@ static int32_t recvExact( const MQTTContext_t * pContext,
             totalBytesRecvd = bytesRecvd;
             receiveError = true;
         }
-        else
+        else if( bytesRecvd > 0 )
         {
+            /* Reset the starting time as we have received some data from the network. */
+            entryTimeMs = getTimeStampMs();
```
Comment (muneebahmed10) [inline source/core_mqtt.c:732]: This name is no longer accurate, something like `lastByteTimeMs` would be better (also since we know at least two bytes were read before this function was called, so that will also be the last byte timestamp at time of entry)

File: source/core_mqtt.c [hunk 2]
```diff
@@ -796,11 +808,7 @@ static MQTTStatus_t discardPacket( const MQTTContext_t * pContext,
             elapsedTimeMs = calculateElapsedTime( getTimeStampMs(), entryTimeMs );
 
             /* Update remaining time and check for timeout. */
```
Comment (muneebahmed10) [inline source/core_mqtt.c:810]: This isn't accurate any more

File: source/include/core_mqtt_config_defaults.h [hunk 3]
```diff
@@ -108,6 +108,27 @@
     #define MQTT_PINGRESP_TIMEOUT_MS    ( 500U )
 #endif
 
+/**
+ * @brief The maximum duration of receiving no data over network when
+ * attempting to read an incoming MQTT packet by the #MQTT_ProcessLoop or
+ * #MQTT_ReceiveLoop API functions.
```
Comment (muneebahmed10) [inline source/include/core_mqtt_config_defaults.h:114]:
```suggestion
 * @brief The maximum duration between non-empty network reads while
 * receiving an MQTT packet via the #MQTT_ProcessLoop or #MQTT_ReceiveLoop
 * API functions.
```

File: source/include/core_mqtt_config_defaults.h [hunk 4]
```diff
@@ -108,6 +108,27 @@
...
+ *
+ * When an incoming MQTT packet is detected, the transport receive function
+ * may be called multiple times until all the expected number of bytes for the
+ * packet are received. This timeout represents the maximum duration of polling
+ * for any data to be received over the network for the incoming.
+ * If the timeout expires, the #MQTT_ProcessLoop or #MQTT_ReceiveLoop functions
+ * return #MQTTRecvFailed.
```
Comment (muneebahmed10) [inline source/include/core_mqtt_config_defaults.h:121]:
```suggestion
 * When an incoming MQTT packet is detected, the transport receive function
 * may be called multiple times until all of the expected number of bytes of the
 * packet are received. This timeout represents the maximum polling duration that
 * is allowed without any data reception from the network for the incoming packet.
 * If the timeout expires, the #MQTT_ProcessLoop and #MQTT_ReceiveLoop functions
 * return #MQTTRecvFailed.
```

File: source/core_mqtt.c [hunk 5]
```diff
@@ -71,17 +71,25 @@ static uint32_t calculateElapsedTime( uint32_t later,
...
  *
  * @param[in] pContext Initialized MQTT Context.
  * @param[in] bytesToRecv Number of bytes to receive.
- * @param[in] timeoutMs Time remaining to receive the packet.
+ *
+ * @note This operation calls the transport receive function
+ * repeatedly to read bytes from the network until either:
+ * 1. The requested number of bytes @a bytesToRecv are read.
+ *                    OR
+ * 2. There is a timeout of MQTT_RECV_POLLING_TIMEOUT_MS duration
```
Comment (aggarg) [inline source/core_mqtt.c:83]: No data is received for MQTT_RECV_POLLING_TIMEOUT_MS.

File: source/core_mqtt.c [hunk 6]
```diff
@@ -683,13 +691,12 @@ static MQTTPubAckType_t getAckFromPacketType( uint8_t packetType )
 /*-----------------------------------------------------------*/
 
 static int32_t recvExact( const MQTTContext_t * pContext,
```
Comment (aggarg) [inline source/core_mqtt.c:693]:
```c
static int32_t recvExact( const MQTTContext_t * pContext,
                          size_t bytesToRecv )
{
    uint8_t * pIndex = NULL;
    size_t bytesRemaining = bytesToRecv;
    int32_t totalBytesRecvd = 0, bytesRecvd;
    uint32_t timestampBeforeRecvMs = 0U, timeSpentInRecvMs = 0U;
    TransportRecv_t recvFunc = NULL;
    MQTTGetCurrentTimeFunc_t getTimeStampMs = NULL;
    bool receiveError = false;

    assert( pContext != NULL );
    assert( bytesToRecv <= pContext->networkBuffer.size );
    assert( pContext->getTime != NULL );
    assert( pContext->transportInterface.recv != NULL );
    assert( pContext->networkBuffer.pBuffer != NULL );

    pIndex = pContext->networkBuffer.pBuffer;
    recvFunc = pContext->transportInterface.recv;
    getTimeStampMs = pContext->getTime;

    while( ( bytesRemaining > 0U ) && ( receiveError == false ) )
    {
        timestampBeforeRecvMs = getTimeStampMs();

        bytesRecvd = recvFunc( pContext->transportInterface.pNetworkContext,
                               pIndex,
                               bytesRemaining );

        timeSpentInRecvMs = calculateElapsedTime( getTimeStampMs(), timestampBeforeRecvMs );

        if( bytesRecvd < 0 )
        {
            LogError( ( "Network error while receiving packet: ReturnCode=%ld.",
                        ( long int ) bytesRecvd ) );
            totalBytesRecvd = bytesRecvd;
            receiveError = true;
        }
        else if( bytesRecvd > 0 )
        {
            /* It is a bug in the application's transport receive implementation
             * if more bytes than expected are received. To avoid a possible
             * overflow in converting bytesRemaining from unsigned to signed,
             * this assert must exist after the check for bytesRecvd being
             * negative. */
            assert( ( size_t ) bytesRecvd <= bytesRemaining );

            bytesRemaining -= ( size_t ) bytesRecvd;
            totalBytesRecvd += ( int32_t ) bytesRecvd;
            pIndex += bytesRecvd;
            LogDebug( ( "BytesReceived=%ld, BytesRemaining=%lu, "
                        "TotalBytesReceived=%ld.",
                        ( long int ) bytesRecvd,
                        ( unsigned long ) bytesRemaining,
                        ( long int ) totalBytesRecvd ) );
        }
        else
        {
            /* No bytes were read from the network. */
        }

        /* If there is more data to be received and nothing was received
         * for MQTT_RECV_POLLING_TIMEOUT_MS, treat this as error. */
        if( ( bytesRemaining > 0U ) &&
            ( timeSpentInRecvMs >= MQTT_RECV_POLLING_TIMEOUT_MS ) &&
            ( bytesRecvd == 0 ) )
        {
            LogError( ( "Time expired while receiving packet." ) );
            receiveError = true;
        }
    }

    return totalBytesRecvd;
}
```

File: test/unit-test/core_mqtt_utest.c [hunk 7]
```diff
@@ -853,7 +867,7 @@ void test_MQTT_Connect_partial_receive()
 
     setupTransportInterface( &transport );
     setupNetworkBuffer( &networkBuffer );
-    transport.recv = transportRecvOneByte;
+    transport.recv = transportRecvNoData;
```
Comment (aggarg) [inline test/unit-test/core_mqtt_utest.c:870]: Is it possible to write a test to replicate the scenario which caught this issue:

-  revc returns less than requested data (possibly return n/2 bytes so that second call with return complete data). 
- processLoop is called with 0 timeout. 

This test should fail without this change and pass with this change.

File: source/core_mqtt.c (hunk 6 above)
Comment (muneebahmed10) [reply source/core_mqtt.c:693]: This will only calculate the time spent for each call of `recvFunc`. Wouldn't there be an infinite loop if every `recvFunc` call returned 0 bytes but spent slightly less than MQTT_RECV_POLLING_TIMEOUT_MS?

Comment (aggarw13) [reply source/core_mqtt.c:693]: In your code suggestion, `timeSpentInRecvMs` represents the time spent only in a single transport recv call, where as my changes look at the time spent across multiple calls to transport recv function that return zero bytes to determine if there has been a polling timeout. 
With your code suggestion, there can be an issue of getting into an infinite loop of the transport recv function returning zero bytes, but the underlying SOCKET_TIMEOUT of transport recv being < MQTT_RECV_POLLING_TIMEOUT_MS, which will prevent the `receiveError` from being set to `true` and thus, the loop will not terminate

File: source/core_mqtt.c [hunk 8]
```diff
@@ -71,17 +71,25 @@ static uint32_t calculateElapsedTime( uint32_t later,
...
  * @param[in] pContext Initialized MQTT Context.
  * @param[in] bytesToRecv Number of bytes to receive.
- * @param[in] timeoutMs Time remaining to receive the packet.
+ *
+ * @note This operation calls the transport receive function
+ * repeatedly to read bytes from the network until either:
+ * 1. The requested number of bytes @a bytesToRecv are read.
+ *                    OR
+ * 2. There is a timeout of MQTT_RECV_POLLING_TIMEOUT_MS duration
+ * between receiving bytes over the network.
```
Comment (aggarw13) [inline source/core_mqtt.c:84]:
```suggestion
 * 2. No data is received from the network for MQTT_RECV_POLLING_TIMEOUT_MS duration.
 * 
```

File: source/core_mqtt.c [hunk 9]
```diff
@@ -683,13 +691,12 @@ static MQTTPubAckType_t getAckFromPacketType( uint8_t packetType )
...
 static int32_t recvExact( const MQTTContext_t * pContext,
-                          size_t bytesToRecv,
-                          uint32_t timeoutMs )
+                          size_t bytesToRecv )
 {
     uint8_t * pIndex = NULL;
     size_t bytesRemaining = bytesToRecv;
     int32_t totalBytesRecvd = 0, bytesRecvd;
-    uint32_t entryTimeMs = 0U, elapsedTimeMs = 0U;
+    uint32_t entryTimeMs = 0U, noDataRecvdTimeMs = 0U;
```
Comment (aggarw13) [inline source/core_mqtt.c:699]:
```suggestion
    uint32_t lastDataRecvTimeMs = 0U, noDataRecvdTimeMs = 0U;
```

File: source/core_mqtt.c [hunk 10]
```diff
@@ -737,10 +747,13 @@ static int32_t recvExact( const MQTTContext_t * pContext,
                         ( unsigned long ) bytesRemaining,
                         ( long int ) totalBytesRecvd ) );
         }
+        else
+        {
+            /* No bytes were read from the network. */
+            noDataRecvdTimeMs = calculateElapsedTime( getTimeStampMs(), entryTimeMs );
```
Comment (aggarw13) [inline source/core_mqtt.c:753]:
```suggestion
            noDataRecvdTimeMs = calculateElapsedTime( getTimeStampMs(), lastDataRecvTimeMs );
```

File: source/core_mqtt.c (hunk 1 above)
Comment (aggarw13) [inline source/core_mqtt.c:732]:
```suggestion
            lastDataRecvTimeMs = getTimeStampMs();
```

File: source/core_mqtt.c (hunk 2 above)
Comment (aggarw13) [inline source/core_mqtt.c:810]:
```suggestion
            /* Check for timeout. */
```

Comment (aggarw13) [reply source/core_mqtt.c]: Removed the "Update remaining time" phrase

File: source/core_mqtt.c (hunk 6 above)
Comment (aggarg) [reply source/core_mqtt.c:693]: Agree. In that case, changing the variable names would make it a bit more readable:

```c
entryTimeMs  --> lastDataRecvdTimestampMs
noDataRecvdTimeMs  --> timeSinceLastDataWasRecvd
```

File: test/unit-test/core_mqtt_utest.c (hunk 7 above)
Comment (aggarw13) [reply test/unit-test/core_mqtt_utest.c:870]: We can achieve that with the existing mock recv function that returns only byte at a time. 
As the bug being addressed is primarily with "_less than expected data received_" when the "remaining timeout" value is nearing zero, I can add a test to mimic that by passing `0` as the timeout to `MQTT_ProcessLoop` and using the `transportRecvOneByte` mock function

File: source/include/core_mqtt_config_defaults.h [hunk 11]
```diff
@@ -108,6 +108,27 @@
...
+ * is allowed without any data reception from the network for the incoming packet.
+ * If the timeout expires, the #MQTT_ProcessLoop and #MQTT_ReceiveLoop functions
+ * return #MQTTRecvFailed.
+ *
+ * <b>Possible values:</b> Any positive integer up to SIZE_MAX. Recommended to
+ * use a small timeout value. <br>
+ * <b>Default value:</b> `10`
+ *
+ */
+#ifndef MQTT_RECV_POLLING_TIMEOUT_MS
```
Comment (sarenameas) [inline source/include/core_mqtt_config_defaults.h:128]: Can you add this to pages.dox

File: test/cbmc/include/core_mqtt_config.h [hunk 12]
```diff
@@ -70,4 +70,18 @@ struct NetworkContext
...
+ * #MQTT_ReceiveLoop API functions.
+ *
+ * When an incoming MQTT packet is detected, the transport receive function
+ * may be called multiple times until all the expected number of bytes for the
+ * packet are received. This timeout represents the maximum duration of polling
+ * for any data to be received over the network for the incoming.
+ * If the timeout expires, the #MQTT_ProcessLoop or #MQTT_ReceiveLoop functions
+ * return #MQTTRecvFailed.
+ */
+#define MQTT_RECV_POLLING_TIMEOUT_MS            ( 3U )
```
Comment (sarenameas) [inline test/cbmc/include/core_mqtt_config.h:85]: Try changing this to 2
```suggestion
#define MQTT_RECV_POLLING_TIMEOUT_MS            ( 2U )
```

File: test/unit-test/core_mqtt_utest.c (hunk 7 above)
Comment (aggarw13) [reply test/unit-test/core_mqtt_utest.c:872]: Added the test

File: source/include/core_mqtt_config_defaults.h (hunk 11 above)
Comment (aggarw13) [reply source/include/core_mqtt_config_defaults.h:128]: Added

File: source/include/core_mqtt.h [hunk 13]
```diff
@@ -603,6 +603,13 @@ MQTTStatus_t MQTT_Disconnect( MQTTContext_t * pContext );
  * @param[in] timeoutMs Minimum time in milliseconds that the receive loop will
  * run, unless an error occurs.
  *
+ * @note Calling this function blocks the calling context for a time period that
+ * depends on the passed @p timeoutMs, the configured macro, #MQTT_RECV_POLLING_TIMEOUT_MS
+ * and the underlying transport interface implementation timeouts.
```
Comment (leegeth) [inline source/include/core_mqtt.h:608]: I think we need to also mention that if not hitting an error.

Comment (aggarw13) [reply source/include/core_mqtt.h:608]: Added

File: docs/doxygen/pages.dox [hunk 14]
```diff
@@ -188,6 +188,9 @@ Some configuration settings are C pre-processor constants, and some are function
 @section MQTT_PINGRESP_TIMEOUT_MS
 @copydoc MQTT_PINGRESP_TIMEOUT_MS
 
+@section MQTT_RECV_POLLING_TIMEOUT_MS
+@copydoc MQTT_RECV_POLLING_TIMEOUT_MS
```
Comment (yourslab) [inline docs/doxygen/pages.dox:192]: In the doc for `MQTT_Init`, we state the following:
```
If there is no time implementation, it is the responsibility of the application to provide a dummy function to always return 0
```
We need to update this to also that `MQTT_RECV_POLLING_TIMEOUT_MS` must also be set to 0 to prevent an infinite loop.

However, this is now insufficient as one iteration of `MQTT_ReceiveLoop` or `MQTT_ProcessLoop` may not be enough to receive the entire payload. The way around that is to set `MQTT_RECV_POLLING_TIMEOUT_MS` to a desired number of iterations, then pass a timer query function that increments on each call. However, that contradicts labeling the unit of time as milliseconds. However, why does the timeout need to be in milliseconds here anyway?

File: source/core_mqtt.c [hunk 15]
```diff
@@ -737,10 +748,14 @@ static int32_t recvExact( const MQTTContext_t * pContext,
                         ( unsigned long ) bytesRemaining,
                         ( long int ) totalBytesRecvd ) );
         }
+        else
+        {
+            /* No bytes were read from the network. */
+            timeSinceLastDataWasRecvdMs = calculateElapsedTime( getTimeStampMs(), lastDataRecvTimeMs );
```
Comment (yourslab) [inline source/core_mqtt.c:754]: I think expecting that a return value of 0 means no data to receive needs to be documented more clearly.

In POSIX sockets and OpenSSL, for example, a return value of 0 means the server has closed the connection or that a close-notify has been sent. No data to receive will actually return a negative error code. In our transport implementations, that error code is mapped appropriately to 0, but this is not going to be immediately obvious to other engineers who try to make their own transport implementations.

File: docs/doxygen/pages.dox (hunk 14 above)
Comment (aggarw13) [reply docs/doxygen/pages.dox:192]: I think that setting the timer function to a value that returns 0 can also result in an infinite loop for the `MQTT_ProcessLoop` function as this is the condition used for determining that the timeout period has expired:
```
if( elapsedTimeMs > timeoutMs )
{
                break;
}
```
We should change that condition to be `>=` instead...

Comment (aggarw13) [reply docs/doxygen/pages.dox:192]: The goal is to allow unbound number of retries for partial reads of packet data over the network as long as the transport recv function is able to successfully read data over the network. (Reason is that the `recvExact` function is called _after_ some part of the MQTT packet is already read, and thus, we want to attempt to read the remaining packet data as long as the transport recv function does not return an error).
Thus, even if a dummy timer function that always returns 0 is used, as long as the transport receive function is able to perform _partial_ reads, the transport recv calls will be retried.

If a dummy timer that always returns 0 is used and the network read returns 0 bytes, there still wouldn't be an infinite loop inside the `recvExact` function as it uses the `>=` check for the timeout.

File: source/core_mqtt.c (hunk 15 above)
Comment (aggarw13) [reply source/core_mqtt.c:754]: Good point, I have updated that in the transport interface API doc.

File: docs/doxygen/pages.dox (hunk 14 above)
Comment (muneebahmed10) [reply docs/doxygen/pages.dox:192]: > We should change that condition to be >= instead...
That definitely looks like a typo as it is now. MQTT_ReceiveLoop uses `>=`, which looks correct

Comment (aggarw13) [reply docs/doxygen/pages.dox:192]: https://github.com/FreeRTOS/coreMQTT/pull/120/commits/32f029190780689881d7de7db6053a9b01e438ff
Have fixed it for `MQTT_ProcessLoop` in the commit

File: source/core_mqtt.c [hunk 16]
```diff
@@ -71,17 +71,25 @@ static uint32_t calculateElapsedTime( uint32_t later,
...
+ *
+ *                    OR
+ * 3. There is an error in reading from the network.
+ *
  *
  * @return Number of bytes received, or negative number on network error.
  */
 static int32_t recvExact( const MQTTContext_t * pContext,
-                          size_t bytesToRecv,
-                          uint32_t timeoutMs );
```
Comment (muneebahmed10) [inline source/core_mqtt.c:84]: Due to the removal of this parameter, some more timeouts could probably be removed due to similar reasoning. I don't think it's critical though and definitely shouldn't be done in this PR.

File: source/core_mqtt.c [hunk 17]
```diff
@@ -2167,7 +2177,7 @@ MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * pContext,
             elapsedTimeMs = calculateElapsedTime( pContext->getTime(),
                                                   entryTimeMs );
 
-            if( elapsedTimeMs > timeoutMs )
+            if( elapsedTimeMs >= timeoutMs )
```
Comment (muneebahmed10) [inline source/core_mqtt.c:2180]: Thanks

File: source/interface/transport_interface.h [hunk 18]
```diff
@@ -163,7 +163,11 @@ typedef struct NetworkContext NetworkContext_t;
  * @param[in] pBuffer Buffer to receive the data into.
  * @param[in] bytesToRecv Number of bytes requested from the network.
  *
- * @return The number of bytes received or a negative error code.
+ * @return The number of bytes received or a negative value to indicate
+ * error.
+ * @note If no data is available on the network to read and no error
+ * has occurred, zero SHOULD be the return value. Zero should NOT be used
```
Comment (muneebahmed10) [inline source/interface/transport_interface.h:169]: Should or must?

File: source/core_mqtt.c [hunk 19]
```diff
@@ -737,10 +748,14 @@ static int32_t recvExact( const MQTTContext_t * pContext,
...
+        {
+            /* No bytes were read from the network. */
+            timeSinceLastDataWasRecvdMs = calculateElapsedTime( getTimeStampMs(), lastDataRecvTimeMs );
+        }
 
-        elapsedTimeMs = calculateElapsedTime( getTimeStampMs(), entryTimeMs );
-
-        if( ( bytesRemaining > 0U ) && ( elapsedTimeMs >= timeoutMs ) )
+        if( ( bytesRemaining > 0U ) &&
+            ( timeSinceLastDataWasRecvdMs >= MQTT_RECV_POLLING_TIMEOUT_MS ) )
```
Comment (muneebahmed10) [inline source/core_mqtt.c:758]: Nit: I don't really like this name. Not saying it's not accurate, it's just long and I don't think the extra words are very meaningful. For example, something like `timeSinceRecvMs` still conveys the relevant information, but in a more concise manner.

I don't think the variable needs to be this explicit if it's clear from reading the rest of the code. Feel free to disagree.

**Review by muneebahmed10 — APPROVED** [review]

File: source/core_mqtt.c (hunk 19 above)
Comment (aggarw13) [reply source/core_mqtt.c]: Sure, I like the recommendation. Have made the name concise.

**Review by muneebahmed10 — APPROVED** [review]

File: source/interface/transport_interface.h (hunk 18 above)
Comment (aggarw13) [reply source/interface/transport_interface.h]: Changed to MUST

**Review by muneebahmed10 — APPROVED** [review]

File: source/core_mqtt.c (hunk 16 above)
Comment (aggarw13) [reply source/core_mqtt.c:84]: Yes, I agree that the timeout passed to `discardPacket` and the timeout logic in the function can also be removed (as hygiene). However, I decided to keep the scope of this PR limited to just fixing the `recvExact` issue.

**Review by aggarg — APPROVED** [review]

**Review by yourslab — DISMISSED** [review]

File: source/core_mqtt.c [hunk 20]
```diff
@@ -737,10 +748,14 @@ static int32_t recvExact( const MQTTContext_t * pContext,
...
+        else
+        {
+            /* No bytes were read from the network. */
+            timeSinceLastRecvMs = calculateElapsedTime( getTimeStampMs(), lastDataRecvTimeMs );
+        }
 
-        elapsedTimeMs = calculateElapsedTime( getTimeStampMs(), entryTimeMs );
-
-        if( ( bytesRemaining > 0U ) && ( elapsedTimeMs >= timeoutMs ) )
+        if( ( bytesRemaining > 0U ) &&
```
Comment (RichardBarry) [inline source/core_mqtt.c:757]: Should this if(){} be inside the else{} immediately above it so it is only checked if no bytes are received?  That would also prevent timeSinceLastRecvMs being tested while it had its initialisation value of zero (which would only be a problem if MQTT_RECV_POLLING_TIMEOUT_MS was also zero).

Comment (aggarw13) [reply source/core_mqtt.c:757]: Nesting the if() within the above immediate else is ideal, but that increases the complexity score of the `recvExact` function beyond the threshold score of 8 (thereby, causing CI check failure). Keeping the complexity score <=8  is the only reason of not having the if-check nested within the else.

Do we want the behavior of retrying the transport recv call once even when the `MQTT_RECV_POLLING_TIMEOUT_MS` value is set to 0? 
The current behavior will not retry if the `MQTT_RECV_POLLING_TIMEOUT_MS` is set to 0 and the first call to transport recv returned 0 bytes.

Comment (RichardBarry) [reply source/core_mqtt.c:757]: Ref the complexity score - I understand that is the requirement but in this case it is leading to code that is not ideal so perhaps we should reduce complexity in another way?

What do you think to just prohibiting MQTT_RECV_POLLING_TIMEOUT_MS from being 0 with a #error message if it is.  We only go into the recvExact() function when we know there is data to receive - and exiting again without that data will just cause an MQTT reset.  In the case where we are just reading one byte to determine if there is any data we don't want to block - but in this case we already know there is data.

Comment (aggarw13) [reply source/core_mqtt.c]: >Ref the complexity score - I understand that is the requirement but in this case it is leading to code that is not ideal so perhaps we should reduce complexity in another way?

Have reduced the complexity from log message to relocate the if-check in its ideal place :)

**Review by muneebahmed10 — DISMISSED** [review]

File: source/core_mqtt.c (hunk 20 above)
Comment (RichardBarry) [reply source/core_mqtt.c]: Gread - no need to restrict MQTT_RECV_POLLING_TIMEOUT_MS to non zero values.

**Review by muneebahmed10 — APPROVED** [review]

**Review by aggarg — APPROVED** [review]
